File: Python/0129.py
from typing import Optional
#Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        
        def helper(node, path):
            if not node:
                return 0
            new_path = path + [node.val]
            ans = 0
            
            if not node.left and not node.right:
                return int("".join(map(str, new_path)))

            ans += helper(node.left, new_path)
            ans += helper(node.right, new_path)
        
            return ans

        return helper(root, [])

# Topic: Tree, DFS
# Time Complexity: O(n)
# Space Complexity: O(n)

File: C++/0129.cpp

#include <vector>
#include <string>
#include <algorithm>

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int sumNumbers(TreeNode* root) {
        return helper(root, {});
    }
private:
    int helper(TreeNode* node, std::vector<int> path) {
        if (!node) return 0;
        path.push_back(node->val);
        int ans = 0;
        if (!node->left && !node->right) {
            std::string num = "";
            for (int digit : path) {
                num += std::to_string(digit);
            }
            return stoi(num);
        }
        ans += helper(node->left, path);
        ans += helper(node->right, path);
        return ans;
    }
};